pkg/controller/podlocality/controller.go

```diff
@@ -0,0 +1,189 @@
+		}),
+	}, predicate.Funcs{
+		CreateFunc: func(evt event.CreateEvent) bool { return evt.Meta != nil },
+		DeleteFunc: func(evt event.DeleteEvent) bool { return false },
+		UpdateFunc: func(evt event.UpdateEvent) bool {
+			return evt.MetaOld != nil && evt.MetaNew != nil &&
+				(evt.MetaOld.GetLabels()[NodeRegionLabel] != evt.MetaNew.GetLabels()[NodeRegionLabel] ||
+					evt.MetaOld.GetLabels()[NodeZoneLabel] != evt.MetaNew.GetLabels()[NodeZoneLabel])
+		},
+		GenericFunc: func(evt event.GenericEvent) bool { return false },
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hasNode(pod runtime.Object) bool {
+	return pod != nil && pod.(*v1.Pod).Spec.NodeName != ""
+}
+
+func hasSidecarAnnotation(pod runtime.Object) bool {
+	return pod != nil && podHasSidecar(*pod.(*v1.Pod))
+}
+
+func podHasSidecar(pod v1.Pod) bool {
+	return pod.Annotations[IstioSidecarStatusAnnotation] != ""
+}
+
+var _ reconcile.Reconciler = &PodLocalityReconciler{}
+
+// PodLocalityReconciler copies the node's region/zone labels to the pod after it's scheduled to a node
+type PodLocalityReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *PodLocalityReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Processing Pod")
+
+	// Fetch the Pod
+	pod := &v1.Pod{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object
+		return reconcile.Result{}, err
+	}
+
+	if pod.Spec.NodeName == "" {
+		// nothing to do. Pod not scheduled yet.
+		return reconcile.Result{}, nil
+	}
+
+	node := &v1.Node{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Name: pod.Spec.NodeName}, node)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if node.Labels[NodeRegionLabel] == "" && node.Labels[NodeZoneLabel] == "" {
+		reqLogger.Info("The node the pod is scheduled on neither has the region nor zone labels. Nothing to do.")
+		return reconcile.Result{}, nil
+	}
+
+	if pod.Labels[NodeRegionLabel] == node.Labels[NodeRegionLabel] && pod.Labels[NodeZoneLabel] == node.Labels[NodeZoneLabel] {
+		reqLogger.Info("Pod's locality labels match the node's. Nothing to do.")
+		return reconcile.Result{}, nil
+	}
+
+	pod.Labels[NodeRegionLabel] = node.Labels[NodeRegionLabel]
+	pod.Labels[NodeZoneLabel] = node.Labels[NodeZoneLabel]
+
+	err = r.Client.Update(context.TODO(), pod)
```

Comment:
This will sometimes fail trying to update a stale resource. I'm still trying to figure out what's changing the Pod inbetween the `Get()` and `Update()`

Reply:
I suppose the only way to be 100% sure is to implement a Get()-Update()-Retry() loop as described here: https://github.com/kubernetes/client-go/blob/396a06da3bd7e3b0cd19b64a1da70fe1142fbff2/examples/create-update-delete-deployment/main.go#L117

This is a minor issue though, as we're not logging as error:
```
2019-08-09T08:41:52.753Z	INFO	controller_podlocality	podlocality/controller.go:143	Processing Pod	{"Request.Namespace": "test", "Request.Name": "sleep-84d98cdb98-2ktqm"}
2019-08-09T08:41:52.782Z	INFO	controller_podlocality	podlocality/controller.go:183	Error updating pod's labels: Operation cannot be fulfilled on pods "sleep-84d98cdb98-2ktqm": the object has been modified; please apply your changes to the latest version and try again	{"Request.Namespace": "test", "Request.Name": "sleep-84d98cdb98-2ktqm"}
2019-08-09T08:41:53.783Z	INFO	controller_podlocality	podlocality/controller.go:143	Processing Pod	{"Request.Namespace": "test", "Request.Name": "sleep-84d98cdb98-2ktqm"}
2019-08-09T08:41:53.798Z	INFO	controller_podlocality	podlocality/controller.go:187	Successfully added zone and region labels to pod.	{"Request.Namespace": "test", "Request.Name": "sleep-84d98cdb98-2ktqm"}
```

Reply:
Keep in mind that `get()` doesn't go to the API server - it only gets the object from the local cache. `Update()`, on the other hand, does go straight to the API Server (but *doesn't* update the local cache!). 

So, there's always a chance for discrepancies. 

Here, some other controller (most likely the kubelet) is also updating the pod (since this is happening during pod creation), so there's a race between them. Getting a 409 conflict is therefore perfectly normal and needs to be handled properly. 

On the get-update-retry loop: most core k8s controllers just try to retry once. IMO, in our case, there's no need to do that. We return the error and the reconciler gets called again (with proper backoff). This is superior because it doesn't block the reconciler from reconciling other pods. If another pod was enqueued in the meantime, it gets reconciled first, and only then is the first pod reconciled again. The get-update-retry loop in reconcile() itself would prevent other pods from being reconciled while waiting for the backoff delay to expire. 

And there's one other reason why using get-update-retry in our operator doesn't work: the controller-runtime framework we're using directs all reads to the cache. This means that performing a get after a failed update won't help, since 99% of the time, we'll get the same stale version of the object (unless the cache is updated by an incoming watch event in the meantime).

Reply:
Makes sense. Thank you for the explanation!